explain-indexes.md

```diff
@@ -0,0 +1,315 @@
++--------------------------------+---------+-----------+--------------------------------+------------------------------------+
+4 rows in set (0.00 sec)
+
+```
+
+In the above example, the last 20 rows are read from the index `intkey`. These `RowID` values are then retrieved from the table data.
+
+## IndexReader
+
+TiDB supports the _covering index optimization_. If all rows can be retrieved from an index, TiDB will skip the second step that is usually required in an `IndexLookup`. Consider the following two examples:
+
+{{< copyable "sql" >}}
+
+```sql
+EXPLAIN SELECT * FROM t1 WHERE intkey = 123;
+EXPLAIN SELECT id FROM t1 WHERE intkey = 123;
+```
+
+```sql
++-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
+| id                            | estRows | task      | access object                  | operator info                     |
++-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
+| IndexLookUp_10                | 1.00    | root      |                                |                                   |
+| ├─IndexRangeScan_8(Build)     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[123,123], keep order:false |
+| └─TableRowIDScan_9(Probe)     | 1.00    | cop[tikv] | table:t1                       | keep order:false                  |
++-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
+3 rows in set (0.00 sec)
+
++--------------------------+---------+-----------+--------------------------------+-----------------------------------+
+| id                       | estRows | task      | access object                  | operator info                     |
++--------------------------+---------+-----------+--------------------------------+-----------------------------------+
+| Projection_4             | 1.00    | root      |                                | test.t1.id                        |
+| └─IndexReader_6          | 1.00    | root      |                                | index:IndexRangeScan_5            |
+|   └─IndexRangeScan_5     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[123,123], keep order:false |
++--------------------------+---------+-----------+--------------------------------+-----------------------------------+
+3 rows in set (0.00 sec)
+```
+
+Because `id` is also the internal `RowID`, it is stored in the `intkey` index. After using the `intkey` index as part of `└─IndexRangeScan_5`, the value of the `RowID` can be returned directly.
+
+## Point_Get and Batch_Point_Get
+
+The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from the Primary Key. This operator is differs from `IndexLookup` in that all data can be retrieved in a single-step: 
```

Comment:
Not only the primary key. A unique key could satisfy the `Point_Get`.

Reply:
I think this is only the case where the first unique not null key is promoted to the primary key. Otherwise, the unique table row will be on a separate region to the index. I verified this with the following testcase:
```sql
mysql> show create table t1\G
*************************** 1. row ***************************
       Table: t1
Create Table: CREATE TABLE `t1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `intkey` int(11) NOT NULL,
  `pad1` varbinary(1024) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `intkey` (`intkey`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=2000001
1 row in set (0.00 sec)

mysql> ALTER TABLE t1 add unique_key int;
Query OK, 0 rows affected (0.26 sec)

mysql> update t1 set unique_key = id;
Query OK, 1010 rows affected (0.05 sec)
Rows matched: 1010  Changed: 1010  Warnings: 0

mysql> alter table t1 add index (unique_key);
Query OK, 0 rows affected (0.33 sec)

mysql> explain select * from t1 where unique_key = 1234;
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
| id                            | estRows | task      | access object                          | operator info                       |
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
| IndexLookUp_10                | 0.00    | root      |                                        |                                     |
| ├─IndexRangeScan_8(Build)     | 0.00    | cop[tikv] | table:t1, index:unique_key(unique_key) | range:[1234,1234], keep order:false |
| └─TableRowIDScan_9(Probe)     | 0.00    | cop[tikv] | table:t1                               | keep order:false                    |
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
3 rows in set (0.00 sec)
```

@lzmhhh123 can you confirm what you meant if otherwise?

Reply:
@nullnotnil That's right.

Reply:
I think it's buggy after altering table.

```mysql
mysql> CREATE TABLE `t1` (
    ->   `id` int(11) NOT NULL AUTO_INCREMENT,
    ->   `intkey` int(11) NOT NULL,
    ->   `pad1` varbinary(1024) DEFAULT NULL,
    ->   unique_key int,
    ->   PRIMARY KEY (`id`),
    ->   KEY `intkey` (`intkey`),
    ->   UNIQUE KEY (unique_key)
    -> )
    -> ;
Query OK, 0 rows affected (0.00 sec)

mysql> explain select * from t1 where unique_key = 1234;
+-------------+---------+------+----------------------------------------+---------------+
| id          | estRows | task | access object                          | operator info |
+-------------+---------+------+----------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:t1, index:unique_key(unique_key) |               |
+-------------+---------+------+----------------------------------------+---------------+
1 row in set (0.00 sec)
```

Reply:
And here's the condition to use Point Get, https://github.com/pingcap/tidb/blob/adb3a4a42f05e02afbe2667771229640a6434851/planner/core/point_get_plan.go#L645-L651

Reply:
nvm, there is a bug in my testcase :-) I created a non-unique key. I can reproduce:

```sql
mysql> DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE `t1` (
    ->   `id` int(11) NOT NULL AUTO_INCREMENT,
    ->   `intkey` int(11) NOT NULL,
    ->   `pad1` varbinary(1024) DEFAULT NULL,
    ->   PRIMARY KEY (`id`),
    ->   KEY `intkey` (`intkey`)
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=2000001;
Query OK, 0 rows affected (0.01 sec)

mysql> 
mysql> ALTER TABLE t1 ADD unique_key INT;
Query OK, 0 rows affected (0.01 sec)

mysql> UPDATE t1 SET unique_key = id;
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

mysql> ALTER TABLE t1 ADD UNIQUE KEY (unique_key);
Query OK, 0 rows affected (0.01 sec)

mysql> 
mysql> explain select * from t1 where unique_key = 1234;
+-------------+---------+------+----------------------------------------+---------------+
| id          | estRows | task | access object                          | operator info |
+-------------+---------+------+----------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:t1, index:unique_key(unique_key) |               |
+-------------+---------+------+----------------------------------------+---------------+
1 row in set (0.00 sec)
```

Reply:
Thanks! I have improved in this commit: https://github.com/pingcap/docs/pull/3933/commits/39ee9ce055e00c468b29c9a067d11748d61fe978